pkg/insights/resyncer.go

```diff
@@ -249,38 +239,58 @@ func (r *resyncer) Start(stop <-chan struct{}) error {
-			}
-			if desc.Scope == model.ScopeGlobal && desc.Name != core_mesh.MeshType {
-				continue
-			}
-			meshName := resourceChanged.Key.Mesh
-			if desc.Name == core_mesh.MeshType {
-				meshName = resourceChanged.Key.Name
-			}
-			var f actionFlag
-			// 'Update' events doesn't affect MeshInsight except for DataplaneInsight, because that's how we find online/offline Dataplane's status
-			if resourceChanged.Operation != events.Update || resourceChanged.Type == core_mesh.DataplaneInsightType {
-				f |= FlagMesh
+			if triggerEvent, ok := event.(events.TriggerInsightsComputationEvent); ok {
+				ctx := context.Background()
+				r.addMeshesToBatch(ctx, batch, triggerEvent.TenantID, resyncEvents)
+				if err := batch.flush(ctx, resyncEvents); err != nil {
+					log.Error(err, "Flush of batch didn't complete, some insights won't be refreshed until next tick")
+				}
 			}
-			// Only a subset of types influence service insights
-			if resourceChanged.Type == core_mesh.DataplaneType || resourceChanged.Type == core_mesh.DataplaneInsightType || resourceChanged.Type == core_mesh.ExternalServiceType {
-				f |= FlagService
+			if resourceChanged, ok := event.(events.ResourceChangedEvent); ok {
+				desc, err := r.registry.DescriptorFor(resourceChanged.Type)
+				if err != nil {
+					log.Error(err, "Resource is not registered in the registry, ignoring it", "resource", resourceChanged.Type)
+				}
+				if desc.Scope == model.ScopeGlobal && desc.Name != core_mesh.MeshType {
+					continue
+				}
+				meshName := resourceChanged.Key.Mesh
+				if desc.Name == core_mesh.MeshType {
+					meshName = resourceChanged.Key.Name
+				}
+				var f actionFlag
+				// 'Update' events doesn't affect MeshInsight except for DataplaneInsight, because that's how we find online/offline Dataplane's status
+				if resourceChanged.Operation != events.Update || resourceChanged.Type == core_mesh.DataplaneInsightType {
+					f |= FlagMesh
+				}
+				// Only a subset of types influence service insights
+				if resourceChanged.Type == core_mesh.DataplaneType || resourceChanged.Type == core_mesh.DataplaneInsightType || resourceChanged.Type == core_mesh.ExternalServiceType {
+					f |= FlagService
+				}
+				batch.add(r.now(), resourceChanged.TenantID, meshName, f)
 			}
-			batch.add(r.now(), resourceChanged.TenantID, meshName, f)
 		case <-stop:
 			log.Info("stop")
 			return nil
 		}
 	}
 }
 
+func (r *resyncer) addMeshesToBatch(ctx context.Context, batch *eventBatch, tenantID string, resyncEvents chan resyncEvent) {
+	meshList := &core_mesh.MeshResourceList{}
+	tenantCtx := multitenant.WithTenant(ctx, tenantID)
+	if err := r.rm.List(tenantCtx, meshList); err != nil {
+		log.Error(err, "failed to get list of meshes", "tenantId", tenantCtx)
+		return
+	}
+	for _, mesh := range meshList.Items {
+		batch.add(time.Now(), tenantID, mesh.GetMeta().GetName(), FlagMesh|FlagService)
+	}
+	if err := batch.flush(tenantCtx, resyncEvents); err != nil {
```

Comment:
Should it be here? I've noticed that it is duplicated in few times. In case of full resync we are going to resync after each tenantId, and after going through all the tenants

Reply:
It shouldn't! Thanks for catching it